### PR TITLE
Rover: add support for wing sails

### PR DESCRIPTION
**Pull request by IamPete1**
This adds the long promised support for wing sails for rover sailboats. A new servo function (124) wing sail elevator has been added. All (i hope) the relevant places have been update to make sure it does the right thing. 

I have tested that the output does what I expected in SITL, I do not have a boat with a wing sail to test this with, We could add a wing sail model to the sailboat SITL but it would be a fair bit of work.

I think this control method will work to get us started, possibly once some more experience has been gained in the wild with wing sails it can be improved upon.

### CODE REVIEW

File: libraries/SRV_Channel/SRV_Channel.h [hunk 1]
```diff
@@ -147,6 +147,7 @@ class SRV_Channel {
         k_LED_neopixel2         = 121,
         k_LED_neopixel3         = 122,
         k_LED_neopixel4         = 123,
+        K_wingsail_elevator     = 124,
```
Comment (rmackay9) [inline libraries/SRV_Channel/SRV_Channel.h]: little typo on the case of the "K".  should be a small "k"

File: libraries/SRV_Channel/SRV_Channel.cpp [hunk 2]
```diff
@@ -65,7 +65,7 @@ const AP_Param::GroupInfo SRV_Channel::var_info[] = {
...
...nt_tilt,8:mount_roll,9:mount_open,10:camera_trigger,11:release,12:mount2_pan,13:mount2_tilt,14:mount2_roll,15:mount2_open,16:DifferentialSpoilerLeft1,17:DifferentialSpoilerRight1,19:Elevator,21:Rudder,24:FlaperonLeft,25:FlaperonRight,26:GroundSteering,27:Parachute,28:EPM,29:LandingGear,30:EngineRunEnable,31:HeliRSC,32:HeliTailRSC,33:Motor1,34:Motor2,35:Motor3,36:Motor4,37:Motor5,38:Motor6,39:Motor7,40:Motor8,41:MotorTilt,51:RCIN1,52:RCIN2,53:RCIN3,54:RCIN4,55:RCIN5,56:RCIN6,57:RCIN7,58:RCIN8,59:RCIN9,60:RCIN10,61:RCIN11,62:RCIN12,63:RCIN13,64:RCIN14,65:RCIN15,66:RCIN16,67:Ignition,68:Choke,69:Starter,70:Throttle,71:TrackerYaw,72:TrackerPitch,73:ThrottleLeft,74:ThrottleRight,75:tiltMotorLeft,76:tiltMotorRight,77:ElevonLeft,78:ElevonRight,79:VTailLeft,80:VTailRight,81:BoostThrottle,82:Motor9,83:Motor10,84:Motor11,85:Motor12,86:DifferentialSpoilerLeft2,87:DifferentialSpoilerRight2,88:Winch,89:Main Sail,90:Camera ISO,91:Camera Focus,92:Camera Shutter Speed,94:Script 1,95:Script 2,96:Script 3,97:Script 4,98:Script 5,99:Script 6,100:Script 7,101:Script 8,102:Script 9,103:Script 10,104:Script 11,105:Script 12,106:Script 13,107:Script 14,108:Script 15,109:Script 16, 124: wing sail elevator
```
Comment (rmackay9) [inline libraries/SRV_Channel/SRV_Channel.cpp]: let's make it "124:Wing Sail Elevator" or "124:WingSailElevator" to be more consistent with the case of other entries

File: APMrover2/sailboat.cpp [hunk 3]
```diff
@@ -211,11 +224,26 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
...
-
         mainsail_out = constrain_float((mainsail_base + pid_offset), 0.0f ,100.0f);
     }
+
+    //
+    // wingsail control
+    //
+
+    // wing sails auto trim, we only need to reduce power if we are tipping over
+    wingsail_out = 100.0f - pid_offset;
```
Comment (rmackay9) [inline APMrover2/sailboat.cpp]: Does this mean that the wing sail is nearly always fully left or right?  I'm probably misunderstanding something..

**Comment by rmackay9** [comment]
Really great addition!

File: APMrover2/sailboat.cpp (hunk 3 above)
Comment (IamPete1) [reply APMrover2/sailboat.cpp]: yeah, it is for use with a sail such as [this ](https://www.wired.com/images_blogs/autopia/2014/02/ff_saildrone_f.jpg) full elevator gives full power either tack, there is no need to trim it for wind direction.

Comment (IamPete1) [reply APMrover2/sailboat.cpp]: it could be that for sailing upwind less elevator is more efficient, but this is probably to nuanced to implement until we do some proper testing

Comment (rmackay9) [reply APMrover2/sailboat.cpp]: The elevator is really there to aim the sail right?  It's not the wagging of the sail that provides the thrust of course?  If the elevator is for aiming the sail then I think the elevator output should reach an equilibrium and that would mostly not be full left or right always.

Comment (IamPete1) [reply APMrover2/sailboat.cpp]: exactly, the elevator just points the sail. The idea is that the user would set the servo min and max to a good equilibrium point for sailing on either tack.

File: libraries/SRV_Channel/SRV_Channel.cpp (hunk 2 above)
Comment (IamPete1) [reply libraries/SRV_Channel/SRV_Channel.cpp]: fixed

File: libraries/SRV_Channel/SRV_Channel.h (hunk 1 above)
Comment (IamPete1) [reply libraries/SRV_Channel/SRV_Channel.h]: fixed

**Comment by IamPete1** [comment]
rebased

**Comment by IamPete1** [comment]
Rebased and fixed a manual mode bug found by a user testing this. This also needs a fix for https://github.com/ArduPilot/ardupilot/issues/12744 before it can be used

**Comment by IamPete1** [comment]
@rmackay9 more stuff left over from GSoC, this is safe now we have more servo functions, has been tested by Karl although not extensively. As I said at the top: 'I think this control method will work to get us started, possibly once some more experience has been gained in the wild with wing sails it can be improved upon.'

File: APMrover2/sailboat.cpp [hunk 4]
```diff
@@ -207,11 +208,23 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
...
+
+    // use PID controller to sheet out
+    float pid_offset = rover.g2.attitude_control.get_sail_out_from_heel(radians(sail_heel_angle_max), rover.G_Dt) * 100.0f;
+    pid_offset = constrain_float(pid_offset, 0.0f, 100.0f);
+
     //
     // mainsail control
     //
-    // if we are motoring or attempting to reverse relax the sail
-    if (motor_state == UseMotor::USE_MOTOR_ALWAYS || !is_positive(desired_speed)) {
```
Comment (rmackay9) [inline APMrover2/sailboat.cpp]: Is the removal of the motor_state and then lower down the pid_offset intentional?  These two changes seem unrelated..

File: APMrover2/sailboat.cpp (hunk 3 above)
Comment (rmackay9) [reply APMrover2/sailboat.cpp]: I also strongly suspect this will need to be changed after a bit of testing.  Surely it's going to stall or turn the sail too far so the wind hits the back of the wing or something like this.  Anyway, I won't hold up the PR for this 'cuz as you say, it's a start.

**Comment by rmackay9** [comment]
As mentioned this look pretty OK, just a couple of changes that I think might be related to a bad rebase.

File: APMrover2/sailboat.cpp (hunk 4 above)
Comment (IamPete1) [reply APMrover2/sailboat.cpp]: They have been moved up,  motor_state  is on line 212 and the PID offset has been moved so we can use it for both the mainsail and wingsail. The reason for this re-jig is that we can go backwards with a wing sail (no promising it will work well, but it is possible) but not a 'normal' sail.  Thanks for taking a look

**Comment by rmackay9** [comment]
Merged, thanks!
